src/main/java/org/openjdk/jextract/impl/CDeclarationPrinter.java

```diff
@@ -0,0 +1,261 @@
+        builder.append(SPACES.substring(0, align));
+    }
+
+    private final StringBuilder builder = new StringBuilder();
+
+    private String print(Declaration decl) {
+        decl.accept(this, null);
+        return builder.toString();
+    }
+
+    // Return C source style signature for the given declaration.
+    // The prefix is emitted for every line. This can be used
+    // to prefix per line comment character "*" in generated javadoc.
+    static String declaration(Declaration decl, String prefix) {
+        Objects.requireNonNull(decl);
+        Objects.requireNonNull(prefix);
+        return new CDeclarationPrinter(prefix).print(decl);
+    }
+
+    static String declaration(Type.Function funcType, String name) {
+        return nameAndType(funcType, "*" + name);
+    }
+
+    @Override
+    public Void visitScoped(Declaration.Scoped d, Void ignored) {
+        indent();
+        var tag = typeTag(d);
+        if (!tag.isEmpty()) {
+            builder.append(tag);
+            if (!d.name().isEmpty()) {
+                builder.append(" " + d.name());
+            }
+            builder.append(" {");
+            builder.append("\n");
+            incr();
+        }
+        d.members().forEach(m -> m.accept(this, null));
+        if (!tag.isEmpty()) {
+            decr();
+            indent();
+            builder.append("};\n");
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitFunction(Declaration.Function d, Void ignored) {
+        indent();
+        var ftype = d.type();
+        var rtype = ftype.returnType();
+        builder.append(nameAndType(rtype, ""));
+        builder.append(" ");
+        builder.append(d.name());
+        builder.append("(");
+        builder.append(
+            d.parameters().
+                stream().
+                map(p -> nameAndType(p.type(), p.name())).
+                collect(Collectors.joining(", "))
+        );
+        if (d.type().varargs()) {
+            builder.append(",...");
+        }
+        builder.append(");\n");
+        return null;
+    }
+
+    @Override
+    public Void visitVariable(Declaration.Variable d, Void ignored) {
+        indent();
+        builder.append(nameAndType(d.type(), d.name()));
+        builder.append(";\n");
+        return null;
+    }
+
+    @Override
+    public Void visitConstant(Declaration.Constant d, Void ignored) {
+        indent();
+        builder.append(nameAndType(d.type(), d.name()));
+        builder.append(" = ");
+        Object value = d.value();
+        if (value instanceof String str) {
+            builder.append("\"" + Utils.quote(str) + "\"");
+        } else {
+            builder.append(value);
+        }
+        builder.append(";\n");
+        return null;
+    }
+
+    @Override
+    public Void visitTypedef(Declaration.Typedef d, Void ignored) {
+        indent();
+        builder.append("typedef ");
+        builder.append(nameAndType(d.type(), d.name()));
```

Comment:
This use of `nameAndType` also leaves me a bit doubtful. There's var declarations, and there's just types + name. What's acceptable in both cases is different - e.g. `int a[3]` is not a valid type for a typedef. So, in cases where we just want a type, and the type should not have a name, we should have a way to say so in the code.

Reply:
Example:

typedef int (*handler)(int x);

The type signature has buried name here ("handler")